### PR TITLE
added MustRunInProc in IBuiltInAnalyzer and some clean up around serializing Solution/Project/DocumentId and how DocumentState is exposed.

### PR DESCRIPTION
**Pull request by heejaechang**
actual change is in https://github.com/dotnet/roslyn/pull/12762/commits/60e9b84414ef0cc9485006145e1c06de115f5f9e

3 mixed changes.

1. IBuiltInAnalyzer.MustRunInProc
this tells diagnostic system whether this particular builtin diagnostic analyzer can be run out of proc.

2. some clean up around Solution/Project/DocumentId.CreateFromSerialized API

3. some clean up on how DocumentState is exposed from Document.

### CODE REVIEW

**Comment by heejaechang** [comment]
@dotnet/roslyn-ide @mattwar @mavasani can you take a look?

**Comment by heejaechang** [comment]
@dotnet/roslyn-ide can someone from ide can take a look?

File: src/Features/Core/Portable/Diagnostics/IBuiltInAnalyzer.cs [hunk 1]
```diff
@@ -8,6 +8,14 @@ namespace Microsoft.CodeAnalysis.Diagnostics
...
     {
+        /// <summary>
+        /// This category will be used to run analyzer more efficiently by restricting scope of analysis
+        /// </summary>
         DiagnosticAnalyzerCategory GetAnalyzerCategory();
+
+        /// <summary>
+        /// This indicates whether this builtin analyzer must run in proc or can be run on remote host such as service hub
+        /// </summary>
+        bool MustRunInProc { get; }
```
Comment (mavasani) [inline src/Features/Core/Portable/Diagnostics/IBuiltInAnalyzer.cs]: Probably a better name and doc comment that helps the analyzer make a call on what value it should provide, say "NeedsWorkspaceContext" or something?

**Comment by heejaechang** [comment]
@mavasani updated comments.

File: src/Features/Core/Portable/Diagnostics/IBuiltInAnalyzer.cs [hunk 2]
```diff
@@ -8,6 +8,20 @@ namespace Microsoft.CodeAnalysis.Diagnostics
...
+        /// <summary>
+        /// This indicates whether this builtin analyzer must run in proc or can be run on remote host such as service hub.
+        /// 
+        /// if the diagnostic analyzer can run in command line as it is, then it should be able to run in remote host. 
+        /// otherwise, it won't unless diagnostic analyzer author make changes in remote host to provide whatever missing
+        /// data command line build doesn't provide such as workspace options/services/MEF and etc.
+        /// 
+        /// at this moment, remote host provide same context as command line build and only that context
+        /// </summary>
+        bool MustRunInProc { get; }
```
Comment (mavasani) [inline src/Features/Core/Portable/Diagnostics/IBuiltInAnalyzer.cs]: LGTM, though I would still prefer a more appropriate property name. IDE folks can decide what works for them..

Comment (heejaechang) [inline src/Features/Core/Portable/Diagnostics/IBuiltInAnalyzer.cs]: created https://github.com/dotnet/roslyn/issues/12767

File: src/Features/Core/Portable/Diagnostics/IBuiltInAnalyzer.cs (hunk 1 above)
Comment (heejaechang) [inline src/Features/Core/Portable/Diagnostics/IBuiltInAnalyzer.cs]: I created tracking bug - https://github.com/dotnet/roslyn/issues/12767

**Comment by heejaechang** [comment]
ping?

**Comment by heejaechang** [comment]
can someone take a look at this?

**Comment by heejaechang** [comment]
this PR is merged to this 
https://github.com/dotnet/roslyn/pull/12774

**Pull request closed without merging**